### ISSUE TITLE
[remote] "code" command spawns multiple blank windows instead of opening a file when called from WSL

### ISSUE THREAD

**Issue by KyleRAnderson** [microsoft/vscode#136106]
Issue Type: <b>Bug</b>

When using the "code" command from a WSL distro's terminal and when passing a path to a file, instead of opening the file, multiple empty Visual Studio Code windows just start popping up.

This does not happen when the executable is invoked from a Windows shell.

I have tried this on two different computers and the results are the same. The settings are synced between these computers however, but one's on Windows 10 20H2 (the one reported) and the other's on Windows 11 21H2.

I'm also not sure if this is a VSCode bug or a bug on the "Remote - WSL" extension.

1. Install the Remote - WSL extension. This is all you need to have installed.
1. Set "security.workspace.trust.untrustedFiles" to "newWindow" in your user's `settings.json`.
1. Open a terminal in a WSL distro.
1. `touch test.txt`, ensuring that `test.txt` is not trusted by workspace trust.
1. `code test.txt`

https://user-images.githubusercontent.com/22606608/139433760-1ba6f899-4fca-4b87-aef5-45849423b6ce.mp4

VS Code version: Code 1.61.2 (6cba118ac49a1b88332f312a8f67186f7f3c1643, 2021-10-19T14:57:20.575Z)
OS version: Windows_NT x64 10.0.19042
Restricted Mode: No
Remote OS version: Linux x64 5.10.16.3-microsoft-standard-WSL2
Remote OS version: Linux x64 5.10.16.3-microsoft-standard-WSL2
Remote OS version: Linux x64 5.10.16.3-microsoft-standard-WSL2
Remote OS version: Linux x64 5.10.16.3-microsoft-standard-WSL2
Remote OS version: Linux x64 5.10.16.3-microsoft-standard-WSL2

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz (8 x 2112)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: enabled_on<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.88GB (1.06GB free)|
|Process Argv|--crash-reporter-id f1b78eeb-acbd-4eea-9d09-3b3a3b816af2|
|Screen Reader|no|
|VM|21%|

|Item|Value|
|---|---|
|Remote|Dev Container: C++|
|OS|Linux x64 5.10.16.3-microsoft-standard-WSL2|
|CPUs|Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz (8 x 2112)|
|Memory (System)|4.82GB (0.24GB free)|
|VM|0%|

|Item|Value|
|---|---|
|Remote|Dev Container: C++|
|OS|Linux x64 5.10.16.3-microsoft-standard-WSL2|
|CPUs|Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz (8 x 2112)|
|Memory (System)|4.82GB (0.23GB free)|
|VM|0%|

|Item|Value|
|---|---|
|Remote|Dev Container: C++|
|OS|Linux x64 5.10.16.3-microsoft-standard-WSL2|
|CPUs|Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz (8 x 2112)|
|Memory (System)|4.82GB (0.24GB free)|
|VM|0%|

|Item|Value|
|---|---|
|Remote|WSL: Ubuntu|
|OS|Linux x64 5.10.16.3-microsoft-standard-WSL2|
|CPUs|Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz (8 x 2112)|
|Memory (System)|4.82GB (0.23GB free)|
|VM|0%|

|Item|Value|
|---|---|
|Remote|Dev Container: C++|
|OS|Linux x64 5.10.16.3-microsoft-standard-WSL2|
|CPUs|Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz (8 x 2112)|
|Memory (System)|4.82GB (0.24GB free)|
|VM|0%|
</details><details><summary>Extensions (19)</summary>

Extension|Author (truncated)|Version
---|---|---
Bookmarks|ale|13.2.2
jupyter-keymap|ms-|1.0.0
remote-containers|ms-|0.202.5
remote-ssh|ms-|0.65.8
remote-ssh-edit|ms-|0.65.8
remote-wsl|ms-|0.58.4
vscode-remote-extensionpack|ms-|0.21.0
slim|sia|0.1.2
vscode-ruby|win|0.28.0
todo-tree|Gru|0.0.214
search-editor-apply-changes|jak|0.1.1
vscode-docker|ms-|1.17.0
python|ms-|2021.10.1365161279
vscode-pylance|ms-|2021.10.3
jupyter|ms-|2021.9.1101343141
jupyter-keymap|ms-|1.0.0
jupyter-renderers|ms-|1.0.3
cpptools|ms-|1.7.1
vscodeintellicode|Vis|1.2.14

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383:30185418
pythonvspyt602:30300191
vspor879:30202332
vspor708:30202333
vspor363:30204092
pythonvspyt639:30300192
pythontb:30283811
pythonvspyt551:30345470
pythonptprofiler:30281270
vsdfh931:30280409
vshan820:30294714
vstes263cf:30335440
vscoreces:30384385
pythondataviewer:30285071
pythonvsuse255:30340121
vscod805:30301674
pythonvspyt200:30340761
binariesv615:30325510
vsccppwt:30382697
pythonvssor306:30344512
bridge0708:30335490
pygetstartedt3:30385195
dockerwalkthru:30377721
bridge0723:30353136
pythonrunftest32:30373476
pythonf5test824:30373475
javagetstartedt:30364666
pythonvspyt187:30373474
vsqsis200cf:30386380
vsaa593:30376534
vssld246:30386377

```

</details>

<details>
<summary>User settings.json</summary>

```jsonc
{
  "git.autofetch": true,
  "python.linting.pylintArgs": [
    "--extension-pkg-whitelist=cv2"
  ],
  "editor.minimap.enabled": true,
  "todo-tree.highlights.enabled": true,
  "git.confirmSync": false,
  "cSpell.language": "en,en-US",
  "workbench.activityBar.visible": true,
  "workbench.sideBar.location": "left",
  "eslint.lintTask.enable": true,
  "cSpell.userWords": [
    "customizability",
    "downcasing",
    "prepending",
    "unprocessable",
    "urljoin"
  ],
  "cSpell.enableFiletypes": [
    "ruby"
  ],
  "java.semanticHighlighting.enabled": true,
  "java.configuration.checkProjectSettingsExclusions": false,
  "vsintellicode.modify.editor.suggestSelection": "choseToUpdateConfiguration",
  "gitlens.hovers.currentLine.over": "line",
  "gitlens.hovers.enabled": false,
  "python.languageServer": "Pylance",
  "python.showStartPage": false,
  "editor.acceptSuggestionOnEnter": "off",
  "python.analysis.typeCheckingMode": "basic",
  "explorer.confirmDragAndDrop": false,
  "jupyter.interactiveWindowMode": "perFile",
  "C_Cpp.clang_format_fallbackStyle": "Google",
  "C_Cpp.clang_format_style": "Google",
  "remote.containers.defaultExtensions": [
    "gruntfuggly.todo-tree",
    "eamodio.gitlens",
    "alefragnani.bookmarks",
    "visualstudioexptteam.vscodeintellicode"
  ],
  "extensions.ignoreRecommendations": true,
  "[yaml]": {
    "editor.defaultFormatter": "redhat.vscode-yaml"
  },
  "workbench.editorAssociations": {
    "git-rebase-todo": "default",
    "*.ipynb": "jupyter-notebook"
  },
  "remote.SSH.remotePlatform": {
    "wato-tr": "linux",
    "rhoderson3": "linux",
    "uwcs": "linux",
    "ubuntu-delta": "linux",
    "ubuntu-thor": "linux",
    "ubuntu-tr": "linux",
    "powerhouse": "linux",
    "dev-vm": "linux",
    "dev-proxied": "linux",
    "dev-noproxy": "linux",
    "debian10": "linux",
    "aws_interviews": "linux",
    "workstation": "linux"
  },
  "[cpp]": {
    "editor.defaultFormatter": "ms-vscode.cpptools"
  },
  // IntelliCode keeps setting this to first, just go with it?
  "editor.suggestSelection": "first",
  "[markdown]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "settingsSync.ignoredExtensions": [
    "llvm-vs-code-extensions.vscode-clangd"
  ],
  "[jsonc]": {
    "editor.defaultFormatter": "vscode.json-language-features"
  },
  "workbench.startupEditor": "none",
  "[ruby]": {
    "editor.defaultFormatter": "jnbt.vscode-rufo"
  },
  "window.dialogStyle": "custom",
  "debug.internalConsoleOptions": "neverOpen",
  "redhat.telemetry.enabled": true,
  "terminal.integrated.rightClickBehavior": "default",
  "notebook.cellToolbarLocation": {
    "default": "right",
    "jupyter-notebook": "left"
  },
  "security.workspace.trust.untrustedFiles": "newWindow",
  "gitlens.blame.format": "${author}: ${message|50?} ${agoOrDate|14-}",
  "gitlens.currentLine.enabled": false,
  "gitlens.codeLens.enabled": false,
  "gitlens.statusBar.enabled": false,
  "security.workspace.trust.banner": "never",
  "editor.bracketPairColorization.enabled": true,
  "settingsSync.ignoredSettings": [
    "remote.SSH.path"
  ],
  "[json]": {
    "editor.defaultFormatter": "vscode.json-language-features"
  },
  "terminal.integrated.defaultProfile.windows": "Git Bash",
  "diffEditor.renderSideBySide": false,
  "workbench.editor.splitInGroupLayout": "vertical",
  "diffEditor.ignoreTrimWhitespace": false
}
```
</details>

<details>
<summary>WSL remote settings.json</summary>

I don't have any remote settings set:

```jsonc
{
}
```
</details>

**Comment by aeschli** [comment]
Can you check what the content is of `%userprofile%\AppData\Roaming\Code\storage.json`. The `windowsState` section has the windows that are restored.

**Comment by KyleRAnderson** [comment]
Sure, here is that section of the file the contents (with folders renamed to workspace<number>):

```json
"windowsState": {
        "lastActiveWindow": {
            "backupPath": "C:\\Users\\kylea\\AppData\\Roaming\\Code\\Backups\\1635767402995",
            "uiState": {
                "mode": 1,
                "x": 184,
                "y": 28,
                "width": 1026,
                "height": 768
            }
        },
        "openedWindows": [
            {
                "folder": "vscode-remote://dev-container%2B5c5c77736c245c5562756e74755c686f6d655c6b796c65615c7265706f735c6473612d766d7064/workspaces/workspace1",
                "backupPath": "C:\\Users\\kylea\\AppData\\Roaming\\Code\\Backups\\9248158002e088c7e319edf06f334be6",
                "remoteAuthority": "dev-container+5c5c77736c245c5562756e74755c686f6d655c6b796c65615c7265706f735c6473612d766d7064",
                "uiState": {
                    "mode": 1,
                    "x": 753,
                    "y": -1417,
                    "width": 1490,
                    "height": 779
                }
            },
            {
                "folder": "vscode-remote://dev-container%2B5c5c77736c245c5562756e74755c686f6d655c6b796c65615c7265706f735c61632d656e67696e65/workspaces/workspace2",
                "backupPath": "C:\\Users\\kylea\\AppData\\Roaming\\Code\\Backups\\52834b6928a3a4da0e40130e2fd03c6e",
                "remoteAuthority": "dev-container+5c5c77736c245c5562756e74755c686f6d655c6b796c65615c7265706f735c61632d656e67696e65",
                "uiState": {
                    "mode": 1,
                    "x": 879,
                    "y": -910,
                    "width": 1365,
                    "height": 864
                }
            },
            {
                "folder": "vscode-remote://dev-container%2B5c5c77736c245c5562756e74755c686f6d655c6b796c65615c7265706f735c6473612d73797374656d2d706c7567696e/workspaces/workspace3",
                "backupPath": "C:\\Users\\kylea\\AppData\\Roaming\\Code\\Backups\\09d0d1c05441dbc57c678efb5ceb9028",
                "remoteAuthority": "dev-container+5c5c77736c245c5562756e74755c686f6d655c6b796c65615c7265706f735c6473612d73797374656d2d706c7567696e",
                "uiState": {
                    "mode": 1,
                    "x": -277,
                    "y": -1425,
                    "width": 1289,
                    "height": 821
                }
            },
            {
                "folder": "file:///c%3A/Users/kylea/Documents/workspace5",
                "backupPath": "C:\\Users\\kylea\\AppData\\Roaming\\Code\\Backups\\a93d00245c658e5d1c6b762de6631eb8",
                "uiState": {
                    "mode": 1,
                    "x": 198,
                    "y": -1152,
                    "width": 1172,
                    "height": 819
                }
            },
            {
                "folder": "file:///c%3A/Users/kylea/Documents/Windows%20Server%202019%20VM%20Share",
                "backupPath": "C:\\Users\\kylea\\AppData\\Roaming\\Code\\Backups\\b9995ade809624d9a0d3e8dd239974ae",
                "uiState": {
                    "mode": 1,
                    "x": -207,
                    "y": -838,
                    "width": 1537,
                    "height": 771
                }
            },
            {
                "backupPath": "C:\\Users\\kylea\\AppData\\Roaming\\Code\\Backups\\1634828755613",
                "remoteAuthority": "wsl+Ubuntu",
                "uiState": {
                    "mode": 1,
                    "x": -263,
                    "y": -880,
                    "width": 1758,
                    "height": 818
                }
            },
            {
                "folder": "vscode-remote://dev-container%2B5c5c77736c245c5562756e74755c686f6d655c6b796c65615c7265706f735c6473612d636f7265/workspaces/workspace6",
                "backupPath": "C:\\Users\\kylea\\AppData\\Roaming\\Code\\Backups\\da7952320871409648c86309372bf1b1",
                "remoteAuthority": "dev-container+5c5c77736c245c5562756e74755c686f6d655c6b796c65615c7265706f735c6473612d636f7265",
                "uiState": {
                    "mode": 1,
                    "x": -233,
                    "y": -1067,
                    "width": 1232,
                    "height": 804
                }
            },
            {
                "backupPath": "C:\\Users\\kylea\\AppData\\Roaming\\Code\\Backups\\1635767402995",
                "uiState": {
                    "mode": 1,
                    "x": 184,
                    "y": 28,
                    "width": 1026,
                    "height": 768
                }
            }
        ]
    }
```

**Comment by aeschli** [comment]
Sorry, still searching for some clues. Ir realized that `window.restoreWindows`: all does not restore when opened with an input, so it's not a restoring issue.

Can you run `code --verbose --log=trace test.txt`, check what printed on the terminal (should be one file URI)
and then check the `main` log file?  (either in the output view, or in
`%userprofile%\AppData\Roaming\Code\logs`

**Comment by KyleRAnderson** [comment]
No worries @aeschli. I tried out the recommended steps. Below you should find the shell output from the command, as well as log file:
```bash
code --verbose --log=trace test.txt
Invoking: /mnt/c/Users/kylek/AppData/Local/Programs/Microsoft VS Code/Code.exe resources/app/out/cli.js --verbose --log=trace --file-uri=vscode-remote://wsl+Ubuntu/home/kylek/test.txt --remote=wsl+Ubuntu in /mnt/c/Users/kylek/AppData/Local/Programs/Microsoft VS Code
```

[main.log](https://github.com/microsoft/vscode/files/7456168/main.log)

**Comment by aeschli** [comment]
Thanks @KyleRAnderson I think I think I found the issue. The log confirms that the same file is opened 4 times. There's a difference in case (`wsl+Ubuntu` vs `wsl+ubuntu`) for the last 3 calls.

I already made a bug fix in that area for 1.62, so maybe this is already fixed in the latest insiders and upcoming stable.
But I made another pass and found more instances where we do not compare the authority in the proper way.

While it's ok to have difference in case in the authority, I wonder where they come from in your case.

Can you send me the full `storage.json`? Or check for entries with `wsl+ubuntu`? (I guess it's in `openedPathsList`).

By removing these, you probably can workaround the issue until you get a build with the fix.

**Comment by KyleRAnderson** [comment]
Sure, I had a look at the `storage.json` again and sure enough there are mixed cases of `wsl+Ubuntu` and `wsl+ubuntu`. Doing a search, I found that all instances of the all-lowercase version were coming from `openedPathsList.entries` as you suspected. In particular, I noticed the test.txt entry there:
```json
            {
                "fileUri": "vscode-remote://wsl%2Bubuntu/home/kylek/test.txt",
                "label": "/home/kylek/test.txt",
                "remoteAuthority": "wsl+ubuntu"
            }
```

There were then 13 other entries similar to this, some of them for folders and a few for files.
After removing the entries for `test.txt` and launching again, I can also confirm that `wsl+ubuntu` (all lowercase) is what was placed in the remoteAuthority field, and reproduces this issue.

**Comment by KyleRAnderson** [comment]
Some good news I think. I was able to reproduce this on the latest VSCode insiders and I narrowed it down to just one setting that would determine whether or not this would happen: `security.workspace.trust.untrustedFiles`.

I disabled all extensions except "Remote WSL" and with the following `settings.json`, I was able to reproduce this issue:
```json
{
  "security.workspace.trust.untrustedFiles": "newWindow",
}
```

If the value of this setting is set to anything else, the issue doesn't occur.
I've updated the steps to reproduce for this issue. I guess the key point is that `~/test.txt` has to be an untrusted file, which with an empty `setting.json`, it will be.
